src/main/java/com/github/dockerjava/netty/handler/HttpResponseStreamHandler.java

```diff
@@ -19,45 +16,87 @@
  */
 public class HttpResponseStreamHandler extends SimpleChannelInboundHandler<ByteBuf> {
 
-    private HttpResponseInputStream stream = new HttpResponseInputStream();
+    private ResultCallback<InputStream> resultCallback;
+
+    private final HttpResponseInputStream stream = new HttpResponseInputStream();
 
     public HttpResponseStreamHandler(ResultCallback<InputStream> resultCallback) {
-        resultCallback.onNext(stream);
+        this.resultCallback = resultCallback;
     }
 
     @Override
     protected void channelRead0(ChannelHandlerContext ctx, ByteBuf msg) throws Exception {
+        invokeCallbackOnFirstRead();
+
         stream.write(msg.copy());
     }
 
+    private void invokeCallbackOnFirstRead() {
+        if (resultCallback != null) {
+            resultCallback.onNext(stream);
+            resultCallback = null;
+        }
+    }
+
     @Override
-    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
-        stream.close();
-        super.channelReadComplete(ctx);
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        stream.writeComplete();
+
+        super.channelInactive(ctx);
     }
 
     public static class HttpResponseInputStream extends InputStream {
 
-        private AtomicBoolean closed = new AtomicBoolean(false);
+        private boolean writeCompleted = false;
 
-        private LinkedTransferQueue<ByteBuf> queue = new LinkedTransferQueue<ByteBuf>();
+        private boolean closed = false;
 
         private ByteBuf current = null;
 
-        public void write(ByteBuf byteBuf) {
-            queue.put(byteBuf);
+        private final Object lock = new Object();
+
+        public void write(ByteBuf byteBuf) throws InterruptedException {
+            synchronized (lock) {
+                if (closed) {
+                    return;
+                }
+                while (current != null) {
+                    lock.wait();
```

Comment:
This blocks the call to `write` until the `current` buffer is read completely, right?